src/data/finset.lean

```diff
@@ -140,6 +143,12 @@ exists_mem_of_ne_zero (mt val_eq_zero.1 h)
 
 @[simp] lemma coe_empty : ↑(∅ : finset α) = (∅ : set α) := rfl
 
+lemma coe_nonempty_iff_ne_empty (s : finset α) : nonempty (↑s : set α) ↔ s ≠ ∅  :=
```

Comment:
Here is a general question: should the library favor `nonempty (↑s : set α)` or `s ≠ ∅`? The library should be consistent about using one or the other to state theorems. The second seems more natural and useful to me. If we go with that, then this should be a simplifier rule. If we go the other way, we should probably simplify the second to the first.

Either way, this lemma should be named `nonempty_iff_ne_empty` or `nonempty_coe_iff_ne_empty`. With the current name, it will be very hard to find using tab completion.

@ChrisHughes24 @jcommelin @johoelzl I am curious to hear your thoughts on this.

Reply:
I think the second looks nicer. Otoh, `nonempty` is a class, which might allow the type class system to take care of some proof obligations.

Reply:
The second looks more natural, but is less useful in practice. I noticed that in my proofs I preferred the `nonempty`, because I can immediately extract an element from that with `cases`. But I'm ok with using `s ≠ ∅` as default since that is mathematically more accepted.

I used this name to be consistent with `set.coe_nonempty_iff_ne_empty` in `set.basic`. I would personally also name it `nonempty_iff_ne_empty`.

Reply:
The ability to uses `cases` with `nonempty` is compelling to me, so I am happy to make that the canonical idiom if nobody objects. In that case, we should try (at least over time) to state theorems that way, and also in set.

But I really do think it is better to have the name start with `nonempty`, also for the one in `set`.

Reply:
The general mathlib preference is for the `s ≠ ∅` idiom, although I can see the advantage of an inductive predicate so that you can use `cases` directly. It's still not that much better since you have to case twice; the additional coercion in the middle makes it less desirable since that makes it internally more complicated. I agree that we should use `ne_empty` and `nonempty` consistently to mean `s ≠ ∅` and `nonempty s` respectively, though.

Reply:
Also, I think `nonempty s` reads quite naturally. I can't imagine many mathematicians stumbling over that... (-;